File: hadesc/src/main/kotlin/hadesc/ast/Param.kt
package hadesc.ast

import hadesc.location.HasLocation
import hadesc.location.SourceLocation

data class Param(
    val binder: Binder,
    val annotation: TypeAnnotation?
) : HasLocation {
    override val location: SourceLocation
        get() = SourceLocation.between(binder, annotation ?: binder)
}

data class TypeParam(
    val binder: Binder
) : HasLocation {
    override val location: SourceLocation
        get() = binder.location
}

File: lib/ast/Param.cpp
//
// Created by dhruv on 07/08/20.
//
#include "hades/ast/Param.h"

namespace hades {

Param::Param(SourceLocation location, u8 index, Optional<Identifier> label,
             Optional<const TypeAnnotation *> type) noexcept
    : m_location(location), m_index(index), m_label(label),
      m_type_annotation(type) {}

auto Param::location() const noexcept -> const SourceLocation & {
  return m_location;
}

auto Param::type_annotation() const noexcept
    -> const Optional<const TypeAnnotation *> & {
  return m_type_annotation;
}

} // namespace hades
